### PR TITLE
Rdrp 333

### PR DESCRIPTION
**Pull request by JenCheshire**
RDRP-333: Short form to long form mapping

## Pull Request submission

Updated `create_headcount_cols()` to include `fte_civil`, `fte_defence`, `hc_total` as variables rather than column names.
Updated test for `create_headcount_cols()`.

Created `run_short_to_long()` in `short_to_long.py` to convert relevant short form responses into long form format.  Includes mapping of in-house expenditure, purchases, full-time employment, and headcount total.

Added `run_short_to_long()` to run in `imputation_main.py`.

#### Closes or fixes

Closes RDRP-333
Closes RDRP-327
Closes RDRP-334
Closes RDRP-335  
Closes RDRP-336

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/outputs/short_form_out.py [hunk 1]
```diff
@@ -23,20 +23,22 @@ def create_period_year(df: pd.DataFrame) -> pd.DataFrame:
...
+    based on the percentages of civil and defence in fte_civil (civil)
+    fte_defence (defence). Note that fte_civil and fte_defence measure different
+    things to hc_total so will not in general total to the hc_total value.
 
     Args:
         df (pd.DataFrame): The survey dataframe being prepared for
             short form output.
+        fte_civil (str): Column containing percentage of civil employees.
+        fte_defence (str): Column containing percentage of defence employees.
+        hc_total (str): Column containing total headcount value.
```
Comment (AnneONS) [inline src/outputs/short_form_out.py:44]: It would be nice to add that fte_civil is usually 706, etc

File: src/outputs/short_form_out.py [hunk 2]
```diff
@@ -23,20 +23,22 @@ def create_period_year(df: pd.DataFrame) -> pd.DataFrame:
```
Comment (AnneONS) [inline src/outputs/short_form_out.py:24]: Also in short form output we need to filter for instance 0 only, as we've created the instance 1 and 2 and we don't want every thing in triple!

**Review by AnneONS — COMMENTED** [review]
Brief review- looks good!

File: src/outputs/short_form_out.py (hunk 1 above)
Comment (JenCheshire) [reply src/outputs/short_form_out.py:44]: Docstring updated

File: src/imputation/imputation_main.py [hunk 3]
```diff
@@ -6,6 +6,8 @@
 
 from src.imputation import tmi_imputation as tmi
 from src.imputation.apportionment import run_apportionment
+from src.imputation.short_to_long import run_short_to_long
+from src.outputs.short_form_out import create_headcount_cols
```
Comment (AnneONS) [inline src/imputation/imputation_main.py]: This last one isn't needed in imputation_main.py now as it's called in short_to_long

File: src/imputation/short_to_long.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    civil_df.loc[:, "instance"] = 1
+
+    defence_df.loc[:, "200"] = "D"
+    defence_df.loc[:, "instance"] = 2
+
+    for each in convert_short_to_long:
+        civil_df[each[2]] = civil_df[each[0]]
+        defence_df[each[2]] = defence_df[each[1]]
+
+    df = pd.concat([df, civil_df, defence_df], ignore_index=True)
```
Comment (AnneONS) [inline src/imputation/short_to_long.py]: Could you consider doing this step? (Copied from tmi_imputation.py)
    updated_df = updated_df.sort_values(
        ["reference", "instance"], ascending=[True, True]
    ).reset_index(drop=True)

File: src/imputation/imputation_main.py [hunk 5]
```diff
@@ -34,6 +36,8 @@ def run_imputation(
 
     df = run_apportionment(df)
 
+    df = run_short_to_long(df)
+
     imputed_df, qa_df = tmi.run_tmi(df, keyvars, mapper)
```
Comment (AnneONS) [inline src/imputation/imputation_main.py:41]: Please can you change the line below so that it doesn't just output long forms, and check what you get?

**Review by AnneONS — COMMENTED** [review]
Also check tmi_imputation so that we only apply the duplicates and the imputation on longforms.

File: src/imputation/short_to_long.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        ("706", "707", "Employment total"),
+        ("headcount_civil", "headcount_defence", "headcount_total"),
+    ]
+
+    civil_df = df.loc[
+        df["formtype"] == "0006",
+    ].copy()
+    defence_df = civil_df.copy()
+
+    df.loc[:, "instance"] = 0
```
Comment (JenCheshire) [inline src/imputation/short_to_long.py]: I've added a line so short form instance is set to 0, but not sure if needed

File: src/imputation/tmi_imputation.py [hunk 7]
```diff
@@ -380,7 +381,8 @@ def apply_tmi(df, target_variables, mean_dict):
 def run_tmi(full_df, target_variables, sic_mapper):
     """Function to run imputation end to end and returns the final
     dataframe back to the pipeline"""
-    copy_df = full_df.copy()
+    copy_df = full_df.copy().loc[full_df["formtype"] == "0001", :]
+    filtered_out_df = full_df.copy().loc[full_df["formtype"] != "0001", :]
```
Comment (dpurches) [inline src/imputation/tmi_imputation.py]: Maybe rename these to `shortform_df` and `longform_df`

Comment (JenCheshire) [reply src/imputation/tmi_imputation.py]: Changes made

File: src/imputation/tmi_imputation.py [hunk 8]
```diff
@@ -395,9 +397,14 @@ def run_tmi(full_df, target_variables, sic_mapper):
...
+
+    final_tmi_df.loc[qa_df.index, "211_trim"] = qa_df["211_trim"]
+    final_tmi_df.loc[qa_df.index, "305_trim"] = qa_df["305_trim"]
+
+    df = pd.concat([final_tmi_df, shortform_df])
 
-    final_df.loc[qa_df.index, "211_trim"] = qa_df["211_trim"]
-    final_df.loc[qa_df.index, "305_trim"] = qa_df["305_trim"]
+    final_df = df.sort_values(["reference", "instance"],
+                              ascending=[True, True]).reset_index(drop=True)
```
Comment (dpurches) [inline src/imputation/tmi_imputation.py:403]: @AnneONS Just wanted to check this approach is okay. In `run_tmi` we split `full_df` into two dataframes - one for short forms and one for long forms. We pass the `longform_df` through TMI and then concatenate the result of this and `shortform_df` back together on line 405.

I think previously we just outputted the longforms from imputation so wanted to double check that now outputting both short and long makes sense.

File: src/imputation/short_to_long.py (hunk 6 above)
Comment (AnneONS) [reply src/imputation/short_to_long.py]: Instead, could you fill instance to zero where it is NaN?

File: src/imputation/tmi_imputation.py (hunk 8 above)
Comment (AnneONS) [reply src/imputation/tmi_imputation.py:403]: this looks good

**Review by AnneONS — APPROVED** [review]
all looks good now

**Pull request merged**